### ISSUE TITLE
We can understand the imformation of type, but default function don't work without type annotation 

### ISSUE THREAD

**Issue by kons-9** [rust-lang/rust#98481]
This code works.
So, rust-compilar can understand the type of tree.

```rust
struct Sample {
    tree: BTreeSet<u32>,
}
impl Sample {
    pub fn new() -> Self {
        let mut tree = Default::default();
        Sample { tree }
    }
}
```

I expect that the compilar can also understand the following code.

```rust
struct Sample {
    tree: BTreeSet<u32>,
}
impl Sample {
    pub fn new() -> Self {
        let mut tree = Default::default();
        tree.insert(3);
        Sample { tree }
    }
}
```

However the current output is:

```
error[E0282]: type annotations needed
 --> src/main.rs:8:13
  |
8 |         let mut tree = Default::default();
  |             ^^^^^^^^
  |
  = note: type must be known at this point
help: consider giving `tree` an explicit type
  |
8 |         let mut tree: _ = Default::default();
  |                     +++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `bug` due to previous error
```

Is it bug? or defined behavior?

It is first time to report bug, and it may be strange.
Thank you.

**Comment by Patryk27** [comment]
I think it's expected, since now you've inserted a method call before the type annotation (coming from `Sample { tree }`) and the method invocation (from `Default::default()`).

This should work if you use `let mut tree = Self::default();` 🙂

**Comment by kons-9** [comment]
I'm sorry for the delay in replying, I didn't realize I had receive a reply.
On second thought, as you say, it couldn't infer it because of the method call in it.

Actually, in my project, I was able to do type inference with rust-analizer, so I misunderstood the compiler could do it too. However, I realized that my case is very specific and a lot of code works like rustc. So it may be a bug in rust-analizer.
(However if rustc could improve this I would be very grad.)

Thank you very much.
